tutorials/stats/3_sample_data_provider.py

```diff
@@ -0,0 +1,125 @@
+# %% [markdown]
+"""
+# 3. Sample data provider
+
+This script demonstrates various instrumentation capabilities.
+It also provides data for the dashboard emulating simultaneous queries
+to the service by multiple users.
+
+"""
+
+# %pip install dff[stats]
+
+# %%
+import random
+import asyncio
+from tqdm import tqdm
+from dff.script import Context, Message
+from dff.pipeline import Pipeline, Service, ACTOR, ExtraHandlerRuntimeInfo
+from dff.stats import (
+    default_extractors,
+    OtelInstrumentor,
+)
+from dff.utils.testing import is_interactive_mode
+from dff.utils.testing.toy_script import MULTIFLOW_SCRIPT, MULTIFLOW_REQUEST_OPTIONS
+
+# %%
+# instrumentation code
+dff_instrumentor = OtelInstrumentor.from_url("grpc://localhost:4317", insecure=True)
+dff_instrumentor.instrument()
+
+
+def slot_processor_1(ctx: Context):
+    ctx.misc["slots"] = {"rating": random.randint(1, 10)}
+
+
+def slot_processor_2(ctx: Context):
+    ctx.misc["slots"] = {"current_topic": random.choice(["films", "games", "books", "smalltalk"])}
+
+
+@dff_instrumentor
+async def get_slots(ctx: Context, _, info: ExtraHandlerRuntimeInfo):
+    return ctx.misc["slots"]
+
+
+def confidence_processor(ctx: Context):
+    ctx.framework_states["response_confidence"] = random.random()
+
+
+@dff_instrumentor
+async def get_confidence(ctx: Context, _, info: ExtraHandlerRuntimeInfo):
+    data = {"response_confidence": ctx.framework_states["response_confidence"]}
+    return data
+
+
+# %%
+pipeline = Pipeline.from_dict(
+    {
+        "script": MULTIFLOW_SCRIPT,
+        "start_label": ("root", "start"),
+        "fallback_label": ("root", "fallback"),
+        "components": [
+            Service(slot_processor_1, after_handler=[get_slots]),
+            Service(slot_processor_2, after_handler=[get_slots]),
+            Service(confidence_processor, after_handler=[get_confidence]),
```

Comment:
Even if it doesn't depend on `ACTOR`, I think `confidence_processor` should go after `ACTOR`.

Reply:
Done